src/lib.rs

```diff
@@ -583,126 +893,190 @@ impl CreateBufferMapped<'_> {
 
     /// Unmaps the buffer from host memory and returns a [`Buffer`].
     pub fn finish(self) -> Buffer {
-        backend::device_create_buffer_mapped_finish(self)
-    }
-}
-
-impl Surface {
-    /// Creates a surface from a raw window handle.
-    pub fn create<W: raw_window_handle::HasRawWindowHandle>(window: &W) -> Self {
-        Surface {
-            id: backend::device_create_surface(window),
+        <C as Context>::flush_mapped_data(self.mapped_data, self.detail);
+        Context::buffer_unmap(&*self.context, &self.id);
+        Buffer {
+            context: self.context,
+            id: self.id,
+            //detail: self.detail,
         }
     }
+}
 
-    #[cfg(any(target_os = "ios", target_os = "macos"))]
-    pub fn create_surface_from_core_animation_layer(layer: *mut std::ffi::c_void) -> Self {
-        Surface {
-            id: wgn::wgpu_create_surface_from_metal_layer(layer),
+impl Instance {
+    /// Create an new instance.
+    pub fn new() -> Self {
+        Instance {
+            context: Arc::new(C::init()),
         }
     }
-}
 
-impl Adapter {
     /// Retrieves all available [`Adapter`]s that match the given backends.
     #[cfg(not(target_arch = "wasm32"))]
-    pub fn enumerate(backends: BackendBit) -> Vec<Self> {
-        wgn::wgpu_enumerate_adapters(backends)
+    pub fn enumerate_adapters(&self, backends: wgt::BackendBit) -> Vec<Adapter> {
+        self.context
+            .enumerate_adapters(wgc::instance::AdapterInputs::Mask(backends, || PhantomData))
             .into_iter()
-            .map(|id| Adapter { id })
+            .map(|id| crate::Adapter {
+                id,
+                context: Arc::clone(&self.context),
+            })
             .collect()
     }
 
+    /// Creates a surface from a raw window handle.
+    pub fn create_surface<W: raw_window_handle::HasRawWindowHandle>(&self, window: &W) -> Surface {
+        Surface {
+            id: self.context.instance_create_surface(window),
+        }
+    }
+
+    #[cfg(any(target_os = "ios", target_os = "macos"))]
+    pub fn create_surface_from_core_animation_layer(
```

Comment:
This seems like it should be unsafe (since the layer pointer can forged) and also in an platform extension trait?

Reply:
Yeah, it's the same as for `create_surface`. I'll make those unsafe. It doesn't matter much, since this is internal API.